### PR TITLE
feat(behaviors): Adding global quick tap 

### PR DESCRIPTION
**Pull request by andrewjrae**
Adding an additional option to hold taps to make the quick tap functionality apply to all taps, rather than just the given hold tap. The current implementation uses the key state listener and is perfectly functional. However, I would've liked to have been able to identify any key being tapped (even if it's internal) as well as ignore any time a key was tapped in conjunction with a hold. I looked into this but it seemed more complicated then it was worth since this current implementation covers almost all cases while being much simpler.

So far I've been daily driving this implementation for almost a week and loving it! The fact that it gets rid of almost all input delay when typing is an added bonus I hadn't even considered.

### CODE REVIEW

**Review by okke-formsma — CHANGES_REQUESTED** [review]
I think we can merge the 'global' and 'local' quick-tap behaviors a bit;

* keep the `last_tapped` and `store_last_tapped` names, rename  `last_tapped.tap_deadline` to `timestamp` as you did.
* change the signature of `store_last_tapped` to take a position + timestamp
* store the last tapped key timestamp (either a normal keypress or the hold-tap) in `last_tapped.timestamp`

This would make the global & local behaviors use the same data, and fix a bug in the current quick-tap implementation too. In the current implementation, only hold-tap keys are considered for quick-taps, which I think is a bug;

*    tap hold-tap A
*    tap &kp B
*    tap hold-tap A again

this sequence should never trigger local 'quick-tap', but if the keys are mashed fast enough it will in the existing implementation.

**Review by okke-formsma — CHANGES_REQUESTED** [review]
Looks good. 

I don't love the solution on line 634 but I'm not sure if my suggestion is much better. I'd love to hear some arguments for or against. I'm not opposed to merging as-is.

**Comment by andrewjrae** [comment]
I agree @okke-formsma, the solution to avoid hold events updating the last tapped timestamp is very hack. We briefly discussed what I would consider to be the proper solution on Discord the other day (https://discord.com/channels/719497620560543766/719544230497878116/961355224566624286), but that relies on a larger, future, feature.

If we want to use timestamps to identify the holds I'm fine with that as well, it's definitely a bit less hack, though it will require an extra variable / logic to keep track of the last hold timestamp. I'll leave it up to whoever else wants to review ;).

At the very least I think some comments should likely be added to note the preferred implementation down the line.

File: docs/docs/behaviors/hold-tap.md
Comment (petejohanson) [inline docs/docs/behaviors/hold-tap.md:54]: I don't think regular users will understand what this means. Can we be a bit more verbose, or give some example sequences so this is clear?

**Review by petejohanson — COMMENTED** [review]
Changes make sense, one comment on the documentation.

File: docs/docs/behaviors/hold-tap.md
Comment (caksoylar) [inline docs/docs/behaviors/hold-tap.md:53]:
```suggestion
If global quick tap is enabled, then `quick-tap-ms` will apply not only when the given hold-tap is tapped but for any key tap before it. This effectively disables the hold behavior when typing quickly, which can be quite useful for home row mods. It can also have the effect of removing the input delay when typing quickly.
```

File: docs/docs/behaviors/hold-tap.md
Comment (caksoylar) [inline docs/docs/behaviors/hold-tap.md:55]:
```suggestion
For example, the following hold-tap configuration enables global quick tap with a 125 millisecond term.
```

File: docs/docs/behaviors/hold-tap.md
Comment (caksoylar) [inline docs/docs/behaviors/hold-tap.md:69]:
```suggestion
If you press `&kp A` and then `&gqt LEFT_SHIFT B` **within** 125 ms, then `ab` will be output. Importantly, `b` will be output immediately since it was within the `quick-tap-ms`. This behavior will hold for any key that is output to the computer. The `&gqt LEFT_SHIFT B` binding will only have its underlying hold-tap behavior if it is pressed 125 ms **after** any keycode is sent to the computer.
```

> This behavior will hold for any key that is output to the computer

It's not very clear to me what this means. Does it mean it only applies to `kp` behaviors?

> The `&gqt LEFT_SHIFT B` binding will only have its underlying hold-tap behavior if it is pressed 125 ms **after** any keycode is sent to the computer

I think involving "sent to the computer" might be confusing here, since AFAICT the timing logic is not related to sending HID events. Maybe we can just say "after any key is pressed"  (or released? I am not sure which one when we say "tapped.")

File: docs/docs/behaviors/hold-tap.md
Comment (caksoylar) [inline docs/docs/behaviors/hold-tap.md:71]: This seems to be missing a part.

**Review by caksoylar — COMMENTED** [review]
Some docs questions and nits.

File: docs/docs/behaviors/hold-tap.md
Comment (andrewjrae) [reply docs/docs/behaviors/hold-tap.md]: Yikes, I forgot to save the file, I actually had a different blurb that was supposed to be there.

File: docs/docs/behaviors/hold-tap.md
Comment (andrewjrae) [reply docs/docs/behaviors/hold-tap.md]: Currently the implementation hooks into the keycode state changed event, which is reflective of what actual HID codes are being sent. But I agree that it's likely not the best way to describe the behavior. I've switched to what I hope is a more clear description.

**Review by caksoylar — APPROVED** [review]

**Review by dxmh — APPROVED** [review]
Docs look great to me, thanks!

I believe @okke-formsma and @petejohanson approve of these changes based on their [comments][1] [above][2] and [in Discord][3], but it would be good to have at least one explicit approval of the code on this PR before merging – just to be sure!

[1]: https://github.com/zmkfirmware/zmk/pull/1187#pullrequestreview-936886421
[2]: https://github.com/zmkfirmware/zmk/pull/1187#pullrequestreview-945054356
[3]: https://discord.com/channels/719497620560543766/883452966114324550/968341158566719538

**Review by petejohanson — APPROVED** [review]

**Pull request merged**
